get partitioning endpoint v2 implementation #226:

TebaleloS opened the pull request:
Release notes
* Define and implemented PL/PG SQL function to retrieve partitioning by id.
* Implement the endpoint functionality to retrieve partitioning by id.
* Added Unit and integration tests.

TebaleloS commented:
Release notes:
# Get partitioning endpoint implementation.

salamonpavel commented on server/src/main/scala/za/co/absa/atum/server/api/exception/ServiceError.scala:
```diff
@@ -22,5 +22,6 @@ object ServiceError {
 
   case class GeneralServiceError(message: String) extends ServiceError
   case class ConflictServiceError(message: String) extends ServiceError
+  case class NotFoundException(message: String) extends ServiceError
```
NotFoundServiceError, let's follow the naming convention ...

salamonpavel commented on server/src/main/scala/za/co/absa/atum/server/api/http/BaseEndpoints.scala:
```diff
@@ -77,4 +84,12 @@ trait BaseEndpoints {
     baseEndpoint.in(Api / V2)
   }
 
+  def pathToAPIv1CompatibleFormat(apiURLPath: String): String = {
```
this should be removed

TebaleloS commented:
Release notes
* Define and implemented PL/PG SQL function to retrieve partitioning by id.
* Implement the endpoint functionality to retrieve partitioning by id.
* Added Unit and integration tests.

salamonpavel commented on database/src/main/postgres/runs/V1.9.3__get_partitioning_by_id.sql:
```diff
@@ -0,0 +1,70 @@
+--      41 - Partitioning not found
+--
+-------------------------------------------------------------------------------
+DECLARE
```
no need for the declare section

TebaleloS replied:
Yes.

salamonpavel commented on database/src/test/scala/za/co/absa/atum/database/runs/GetPartitioningByIdIntegrationTests.scala:
```diff
@@ -0,0 +1,95 @@
+      |""".stripMargin
+  )
+
+  private val expectedPartitioning: Json = parse(partitioning.value) match {
```
maybe it could be even simplified since we don't transform the success value in any way
```scala
parse(partitioning.value).getOrElse(throw new Exception("Failed to parse JSON"))
```

salamonpavel commented on database/src/test/scala/za/co/absa/atum/database/runs/GetPartitioningByIdIntegrationTests.scala:
```diff
@@ -0,0 +1,95 @@
+        assert(queryResult.hasNext)
+        val row = queryResult.next()
+        val returnedPartitioning = row.getJsonB("partitioning").get
+        val returnedPartitioningParsed = parse(returnedPartitioning.value).getOrElse(Json.Null)
```
personally I would fail the test rather than using `Json.Null`
```scala
val returnedPartitioningParsed = parse(returnedPartitioning.value).getOrElse(fail("Failed to parse returned partitioning"))
```

salamonpavel commented on server/src/test/scala/za/co/absa/atum/server/api/service/PartitioningServiceUnitTests.scala:
```diff
@@ -139,6 +143,18 @@ object PartitioningServiceUnitTests extends ZIOSpecDefault with TestData {
+            result <- PartitioningService.getPartitioning(1111L)
+          } yield assertTrue(result == partitioningWithIdDTO1)
+        },
+        test("Returns expected ServiceError") {
```
The first test case is OK as you are simulating successful scenario. The other two test cases should be for unexpected error with `ZIO.fail(GeneralDatabaseError("boom!"))` and assert on `GeneralServiceError` and third test case should be for `ZIO.fail(NotFoundDatabaseError("Partitioning not found"))` and assert on `NotFoundServiceError`.

salamonpavel commented on server/src/test/scala/za/co/absa/atum/server/api/controller/PartitioningControllerUnitTests.scala:
```diff
@@ -99,6 +104,20 @@ object PartitioningControllerUnitTests extends ZIOSpecDefault with TestData {
           )
         }
+      ),
+      suite("GetPartitioningSuite")(
```
I would expect three test scenarios. 1. success case (you have that one OK). 2. Case where partitioning was not found, therefore the mock is set up to return NotFoundServiceError and you assert on NotFoundErrorResponse. 3. the unexpected error (you have that one already). You see with the mocks you are supposed to simulate all different scenarios.

salamonpavel commented on server/src/test/scala/za/co/absa/atum/server/api/repository/PartitioningRepositoryUnitTests.scala:
```diff
@@ -174,16 +185,27 @@ object PartitioningRepositoryUnitTests extends ZIOSpecDefault with TestData {
+            result <- PartitioningRepository.getPartitioning(1111L)
+          } yield assertTrue(result == partitioningWithIdDTO1)
+        },
+        test("Returns expected DataNotFoundException") {
```
again, there could/should be actually three test cases

salamonpavel commented on server/src/test/scala/za/co/absa/atum/server/api/repository/PartitioningRepositoryUnitTests.scala:
```diff
@@ -22,11 +22,12 @@ import za.co.absa.atum.server.api.database.runs.functions._
 import zio.interop.catz.asyncInstance
 import zio.test.Assertion.failsWithA
-import zio.test._
+import zio.test.{assertZIO, _}
```
import zio.test._

benedeki commented:
The failing test coverage has bee investigated. It's caused by inaccurate measuring by JaCoco. Nothing to do about it now.
